### DISCUSSION TITLE
Notes Dec 14th, 2022

### DISCUSSION THREAD

**Discussion by wipfli** [maplibre/maplibre#154] [Technical Steering Committee Meetings]
# December 14th, 2022: MapLibre Eastern Call

* 9 AM to 10 AM CET

### Attendees

* Brandon Liu (Protomaps)
* Ian Wagner (Stadia Maps)
* Alan Wei Chen (Grab)
* Jannik (Graphmasters)
* Joao Duarte (one.network)
* Kanahiro Iguchi (MIERUNE)
* Matthew Mamonov
* Yasunori Kirimoto (MIERUNE)
* Taras Vozniuk (Azure maps)
* Yanjin Li (Grab)
* Oliver Wipfli

### Fonts

* Brandon: Font-maker. Almost all MapLibre maps require a font stack. Building the font stack was always a bit exotic. 
    * Should MapLibre distribute a fall-back font?
    * Improve CJK font rendering in MapLibre? Do people complain?
* Ian: Can glyphs can be rendered differently based on where you are?
    * Brandon: It uses locale information and uses your local variants.
* Ian: What about Burmese? Does this work with SDF glyphs?
    * Brandon: The technology does not support Burmese.
* Ian: A Noto-based font could make sense.
* Brandon: Can double pixel resolution to make CJK look better.
* Wei Chen: We have text rendering issues in SE-Asia. Mapbox pre-renders all the text images on the server. Khmer is an interesting language for which we have support in our Mapbox Native fork. We also have and implementation in the web team.
* Ian: Seems to be well solved in general on mobile, but on web it seems a bit harder. Maybe we could use webassembly?
    * Brandon: You would have to use more data (https://github.com/yeslogic/allsorts). Native and web were always looking identical. Should we keep this paradigm?
  

# December 14th, 2022: Native

7 PM to 8 PM CET

### Attendees (First Name, Last Name, GitHub handle, affiliation)

* Steve Gifford @sjg-wdw - Wet Dog Weather, Stamen
* Eugene Zhidkov - [ezhidkov@apple.com](mailto:ezhidkov@apple.com) - Apple Metal team (Eugene cannot speak for Apple until they approve his request to engage with the MapLibre open-source community)
* Oliver Wipfli @wipfli, coordinator at MapLibre
* Tadej Novak @ntadej, independent
* Luke Seelenbinder - @lseelenbinder, board member & Stadia Maps
* Rob C.  @roblabs, representing RobLabs.com 
* Birk Skyum, @birkskyum
* Seth Fitzsimmons, @mojodna, AWS
* Haowen You, AWS
* Prateek, AWS
* Camilla Mahon, Stamen
* Felix Faasen, TomTom
* Maarten Gribnau, TomTom
* Marc Regan, Meta
* Marc Wilson, Meta
* Max Ammann
* Nathan van Fleet, Lyft
* Yuri Astrakhan, MapLibre

Chair: Luke Seelenbinder

### MapLibre GL Native Organizational Questions

* [sjg] Wrapping up [Renderer Modularization PR](https://github.com/maplibre/maplibre-gl-native/pull/547) feedback
    * Modularization prepares the code base for new rendering backends like Metal or Vulkan
    * Got 8 approvals
    * Some open questions from @JannikGM but not all are related to Modularization
    * Would like to discuss interesting ideas with @JannikGM directly in follow-up discussions
    * Can we merge? Luke: Yes.
* [sjg] Starting [Metal Port PR](https://github.com/maplibre/maplibre-gl-native/pull/580) feedback 
    * Proposes a Metal implementation based on the modularization proposal
    * More technical maybe than the rendering modularization.
    * Got review from Marc Wilson and close to an approval.
    * Are two approvals enough? Luke: Maybe 2 approvals from people and 2 approvals from Governing Board members (Luke and Birk volunteer to review).
    * Eugene how long does approval from Apple take for you to be able to contribute to discussions on GitHub? Eugene: usually 1 week.
    * Who else could help? Nathan? Nathan: Do have iOS experience but not so much Metal. Everyone is welcome to review independent of your level of knowledge.
    * Luke: Is it a problem to wait for Eugene? Steve: I could talk with Eugene offline and incorporate his feedback.
    * By Friday we should have the draft finished.
* Luke: What is next with the modularization and Metal?
    * Haowen: Source people to work on the implementation. Can work together with Steve or might also break down into small tasks where some could be covered by Bounties. 
    * Haowen: It is a big effort and we split it already into two parts. From here we want to break down into smaller milestones. I think we can present a more detailed plan in January.
    * Luke: Like the idea of community involvement.
    * Haowen: Metal support ready by the end of 2023 is our high-level goal. And we hope that the knowledge can be kept in the community.
    * Luke: We are hoping to hire by the end of this month for the Maintainer Native role.
* [wipfli] Merged a first pull request from Grab!
    * [https://github.com/maplibre/maplibre-gl-native/pull/581](https://github.com/maplibre/maplibre-gl-native/pull/581)

### MapLibre GL Native Technical Questions

* [wipfli] Making an Android and iOS release
    * [https://github.com/maplibre/maplibre-gl-native/issues/585](https://github.com/maplibre/maplibre-gl-native/issues/585)
    * [https://github.com/maplibre/maplibre-gl-native/issues/584](https://github.com/maplibre/maplibre-gl-native/issues/584)
    * Probably requires self-hosted ubuntu runners due to timeouts
    * Are we good with releasing what we have now? Luke: Yes. Let’s do it.
* [ntadej] Future of Qt Mapping API
    * Qt dropping backend-specific “universal” API
    * Will need MapLibre-specific layer on top of the Qt one
* [roblabs] iOS CI issues.  _Appears_ to be related to file I/O on build machine?!?
    * [https://github.com/maplibre/maplibre-gl-native/pull/599#issuecomment-1351767099](https://github.com/maplibre/maplibre-gl-native/pull/599#issuecomment-1351767099)
    * [https://github.com/maplibre/maplibre-gl-native/pull/594#issuecomment-1351776016](https://github.com/maplibre/maplibre-gl-native/pull/594#issuecomment-1351776016)

### MapLibre Organizational Questions

* [https://github.com/maplibre/font-maker](https://github.com/maplibre/font-maker) is a new repo to convert fonts
* [Max] Introduce an email notification-mechanism for these meetings. Ideas:
    * Newsletter
    * Calendar which allows subscription via email
    * Google Groups?
* [wipfli] Meeting agenda and notes with GitHub logins
    * Google docs is a barrier for people to contribute to agenda
    * We might use [https://hedgedoc.org](https://hedgedoc.org)/ next time (collaborative markdown with GitHub oauth)
* [wipfli] Managing controversial pull requests
    * [https://github.com/maplibre/maplibre/discussions/155](https://github.com/maplibre/maplibre/discussions/155)
    * Can we define recommendations on how people can contribute non-trivial things?
* [wipfli] Meeting time
    * Should we to the Eastern call regularly? Luke: Yes, but the default should be offline and async discussion, because we cannot work around the time-zone problem.
* [roblabs] _Proposal _— we have several different builds on Native.  Would it make sense to have subject matter experts host “Install festivals”?  Goal would be to work through the Markdown docs and make sure anyone can follow Markdown notes to be successful in making any build flavor.  
    * macOS — android, macos, ios, [Tileserver GL](https://github.com/maptiler/tileserver-gl), GitHub self-hosted, [Xcode Server Bots](https://github.com/roblabs/xcode-continuous-integration), bazel, QT, [React](https://github.com/maplibre/maplibre-react-native), [Rust](https://github.com/maplibre/maplibre-rs)
    * Linux — android, tileserver GL, GitHub self-hosted, Rust
    * Prateek: Need good DEVELOPMENT.md files
    * Tadej: Need better cmake and decoupling between core and platform bindings. 
    * Max: If you use only linux, you can use nix. But that does not work with Windows.
* [roblabs] Face to Face meeting opportunities
    * Open Street Map, State of the Map US 2023 — June 8-11, 2023.  Richmond, Virginia, US
    * FOSS4G-North America ??

# December 14th, 2022: Web

8 PM to 9 PM CET

### Attendees (First Name, Last Name, GitHub handle, affiliation)

* Harel Mazor @HarelM - MapLibre/Israel Hiking Map
* Luke Seelenbinder @lseelenbinder - Board Member & Stadia Maps
* Rob C.  @roblabs, representing RobLabs.com
* Birk S. @birkskyum
* Derek Westcott, Microsoft
* David Buerer, Microsoft
* Felix Faasen, TomTom
* Ivan Sanchez
* Petr Pridal, MapLibre
* Matthew Mamonov

Chair: Harel Mazor

### MapLibre GL JS Technical Questions

* Harel: [https://github.com/maplibre/maplibre-gl-js/discussions/1851](https://github.com/maplibre/maplibre-gl-js/discussions/1851) - Change v8.json file to be a typescript definition file
    * Strange that we use a limited language in v8 to define types when we actually could use TypeScript. TypeScript is one option, but you need to be able to share the output with Native / docs. Other option could be JSON-scheme.
    * Luke: In v8 we define types? Harel: It defines the style specification expressions. Now we have types which are for example `number | [number]` under the same property (e.g. icon-padding) but this was very hard to do in the v8 language.
    * Harel: The codebase does not exclusively check types based on the v8.json file. It also has extra code around this.
    * Birk: Mapbox did not build GL JS with a proper type system with the capabilities of TypeScript.
    * Harel: Nobody knows how the language in v8 works. It is documented and not obvious.
    * Harel: Not obvious to use TypeScript style spec for runtime validation. Ship the TypeScript compiler with MapLibre GL JS does not seem like an option.
    * Matthew: Found limitation of current solution when working with multiple sprites in style. Needed array of strings there which is currently not possible.
    * Matthew: How do the native implementations use the v8.json file? Nobody knows.
    * Matthew: Main problem right now is maintainability. We should discuss two things: run-time validation and keeping the proprietary language for v8.json. Harel: Yes let's split the discussion. Find alternative for run-time validation maybe.
    * Birk: The java native part is done like this: A JavaScript script reads the v8.json file and outputs java code.
    * Example usage in native: [https://github.com/maplibre/maplibre-gl-native/blob/main/scripts/generate-style-code.js](https://github.com/maplibre/maplibre-gl-native/blob/main/scripts/generate-style-code.js)
* Harel: [https://github.com/maplibre/maplibre-gl-js/issues/1925](https://github.com/maplibre/maplibre-gl-js/issues/1925) - area in pixels math function addition
    * Is currently possible with existing style expressions
    * Would shorten the expression for that use-case
    * Birk: If we transition to other projections, this would be related
    * Ivan: This could be done server-side and put into the tiles, and you could multiply by zoom level.
    * Oliver: This seems unnecessary since it is already possible with existing features.
    * Birk: It assumes that the label and map is not tilted.
* David: [https://github.com/maplibre/maplibre-gl-js/pull/1800](https://github.com/maplibre/maplibre-gl-js/pull/1800) - Add contentMatch to sprites
    * Saves several 100 ms on startup time
    * Harel: Result of the discussion was that we need better names which are more intuitive. Maybe make 3 to 5 more name suggestions and then we have a vote. I don't have a good suggestion. David: The problem is a bit that it is very maps and shields specific.
    * Harel: The existing sprite code seems to be very specific to shields.
    * Harel: Let's do 2 new name suggestions and then we wrap it up.
* [roblabs] Looking for guidance from the JavaScript world on: _testing on mobile devices._  Are there industry standards for testing GL JS on mobile devices (_ad hoc_, AWS Device Farm, or ____)?  Example:  [#1919](https://github.com/maplibre/maplibre-gl-js/issues/1919)
    * Tyler: We use Cypress and Playwright at AWS. Are you looking for screen resolution or also input like touch?
    * Rob: For example the 2x resolution.
    * Tyler: Run the tests and screenshots in a pipeline. There are mac EC2 now.
    * Harel: Appium has mobile test capabilities.
    * David:  [Visual Studio App Center](https://visualstudio.microsoft.com/app-center/)
    * Ivan:  
        * [https://karma-runner.github.io/](https://karma-runner.github.io/)
        * [https://github.com/leahciMic/karma-android-launcher](https://github.com/leahciMic/karma-android-launcher) 
    * Birk: We have playwright in MapLibre GL JS already and we could do safari testing there. For font rendering and interactivity.
* Tyler: 
    * Removing debug pages
    * Do blog (News) posts support markdown for code sections?

### MapLibre Organizational Questions

* [https://github.com/maplibre/font-maker](https://github.com/maplibre/font-maker) is a new repo to convert fonts
* [Max] Introduce an email notification-mechanism for these meetings. Ideas:
    * Newsletter
    * Calendar which allows subscription via email
    * Google Groups?
* [wipfli] Meeting agenda and notes with GitHub logins
    * Google docs is a barrier for people to contribute to agenda
    * We might use [https://hedgedoc.org](https://hedgedoc.org)/ next time (collaborative markdown with GitHub oauth)
    * Felix: [https://hackmd.io/](https://hackmd.io/)
* [wipfli] Managing controversial pull requests
    * [https://github.com/maplibre/maplibre/discussions/155](https://github.com/maplibre/maplibre/discussions/155)
    * Can we define recommendations on how people can contribute non-trivial things?
    * Tyler: Tenets could be good.
    * Tyler: Do all objections have to be addressed in a pull request? Who can decide if something gets merged?
* [roblabs] MapLibre Plugins for ArcGIS, Carto, WMS, overview & interactivity.  These plugins are designed so that you can just pull in the MapLibre functionality that you need.  Designed by Jim McAndrew 
    * Demo:  [RobLabs.com/nps](https://roblabs.com/nps/) — see this link, and then a further link to videos
* Petr: What is the plan for MapLibre GL JS v3?
    * Harel: Was hoping to get more breaking changes in, but did not get as far. Want to do a pre-release and check the state of plugins.
    * Petr: Webgl 2 could be good, and we have quite some improvements of terrain.
    * Oliver: What is the state with reactive frameworks? Birk: It is a hack at the moment. We could reach out to @Pessimisstress

**Comment by maxammann** [comment]
> What about Burmese? Does this work with SDF glyphs?
Brandon: The technology does not support Burmese.

I just read the Wikipedia article about Brumese. Isn't th only limitation proper text shaping instead of text rendering?

> Brandon: Can double pixel resolution to make CJK look better.

There are also multi-channel SDFs which I want to give a try soon
They claim to be better than SDFs of double the pixel resolution

**Reply by bdon** [reply]
Yes, the current MapLibre GL text rendering system does not implement general shaping apart from specific cases for Hebrew and Arabic via rtl-text-plugin.

**Reply by HarelM** [reply]
Worth mentioning that even for Hebrew (and I believe for Arabic as well) the system has flaws.
For example, if you add "nikud" to Hebrew it will not render property.
The following will work:
שלום
But with nikud:
שָׁלוֹם
It won't work (see the small points and dashes around and inside the letters).
Here's another example of the same problem, I believe:
https://github.com/maplibre/maplibre-gl-js/issues/50
Having said that, I have no clue how to solve this...
